### PR TITLE
WIP: basic tap flow

### PR DESCRIPTION
**Pull request by yemel**

### CODE REVIEW

File: server/src/LightController.js [hunk 1]
```diff
@@ -169,6 +169,10 @@ module.exports = class LightController extends EventEmitter {
     return {defaults, presetOverrides, currentPreset: presetName, overrides};
   }
 
+  tap(cid) {
```
Comment (martinblech) [inline server/src/LightController.js:172]: s/cid/clientId/?

File: server/src/ProgramScheduler.js [hunk 2]
```diff
@@ -91,4 +91,8 @@ module.exports = class ProgramScheduler {
     this.program.updateConfig(config);
   }
 
+  tap(cid) {
```
Comment (martinblech) [inline server/src/ProgramScheduler.js:94]: s/cid/clientId/?

File: server/src/light-programs/base-programs/LightProgram.js [hunk 3]
```diff
@@ -1,5 +1,6 @@
 module.exports = class LightProgram {
   constructor(config, geometry, shapeMapping, lightController) {
+    this.taps = [];
```
Comment (martinblech) [inline server/src/light-programs/base-programs/LightProgram.js:3]: I would leave this instance var out from the base LightProgram class (it's ever growing in size and very unlikely that all concrete programs will need a full history of all taps).

File: server/src/light-programs/base-programs/LightProgram.js [hunk 4]
```diff
@@ -14,6 +15,11 @@ module.exports = class LightProgram {
     throw new Error("Child classes should override drawFrame");
   }
 
+  tap(cid) {
```
Comment (martinblech) [inline server/src/light-programs/base-programs/LightProgram.js:18]: Please leave the base implementation of tap() empty. Subclasses can override to hook in on the event.

Comment (martinblech) [inline server/src/light-programs/base-programs/LightProgram.js:18]: s/cid/clientId/

**Review by martinblech — CHANGES_REQUESTED** [review]

**Comment by yemel** [comment]
Thanks for the review @martinblech.

@catalanojuan included your comments in his PR. I'll close this and continue working based on it.

**Pull request closed without merging**